src/Reporting/src/App.Metrics.Formatting.StatsD/Builder/MetricsStatsDFormatterBuilder.cs

```diff
@@ -0,0 +1,99 @@
+﻿// <copyright file="MetricsStatsDFormatterBuilder.cs" company="App Metrics Contributors">
+// Copyright (c) App Metrics Contributors. All rights reserved.
+// </copyright>
+
+using System;
+
+namespace App.Metrics.Formatting.StatsD.Builder
+{
+    public static class MetricsStatsDFormatterBuilder
+    {
+        /// <summary>
+        ///     Add the <see cref="MetricsStatsDStringOutputFormatter" /> allowing metrics to optionally be reported to Datadog
+        /// </summary>
+        /// <param name="metricFormattingBuilder">
+        ///     The <see cref="IMetricsOutputFormattingBuilder" /> used to configure StatsD formatting
+        ///     options.
+        /// </param>
+        /// <param name="setupAction">The StatsD formatting options to use.</param>
+        /// <param name="fields">The metric fields to report as well as their names.</param>
+        /// <returns>
+        ///     An <see cref="IMetricsBuilder" /> that can be used to further configure App Metrics.
+        /// </returns>
+        public static IMetricsBuilder AsStatsDString(
+            this IMetricsOutputFormattingBuilder metricFormattingBuilder,
+            Action<MetricsStatsDOptions> setupAction,
+            MetricFields fields = null)
+        {
+            if (metricFormattingBuilder == null)
+            {
+                throw new ArgumentNullException(nameof(metricFormattingBuilder));
+            }
+
+            if (setupAction == null)
+            {
+                throw new ArgumentNullException(nameof(setupAction));
+            }
+
+            var options = new MetricsStatsDOptions();
+
+            setupAction(options);
+
+            var formatter = new MetricsStatsDStringOutputFormatter(options, fields);
+
+            return metricFormattingBuilder.Using(formatter, false);
+        }
+
+        /// <summary>
+        ///     Add the <see cref="MetricsStatsDStringOutputFormatter" /> allowing metrics to optionally be reported to Datadog
+        /// </summary>
+        /// <param name="metricFormattingBuilder">
+        ///     The <see cref="IMetricsOutputFormattingBuilder" /> used to configure Datadog formatting
+        ///     options.
+        /// </param>
+        /// <param name="options">The Datadog formatting options to use.</param>
```

Comment:
Nitpick: StatsD is an OSS standard, but DataDog supports it as an interface.

Reply:
I'm having trouble finding the exact DogStatsD wire format. I think DogStatsD added tagging and events, I've added the tag support to this implementation (by appending `|#[tag1]:[value1],...[tagN]:[valueN]` at the end), but I haven't checked their implementation of event sending through StatsD.

Maybe we should differentiate the writer into more specialized writers instead of implementing a generic format writer? The default should work with vanilla StatsD, and have a dedicated writer for DogStatsD, and maybe add other formats later?

Reply:
Note that this tagging format is for AWS StatsD format, which might be the same as the one DogStatsD is using.